components/playlists/browser/playlists_media_file_controller.cc

```diff
@@ -0,0 +1,342 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/playlists/browser/playlists_media_file_controller.h"
+
+#include <utility>
+
+#include "base/bind.h"
+#include "base/files/file.h"
+#include "base/files/file_util.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/task/post_task.h"
+#include "base/task_runner_util.h"
+#include "brave/components/playlists/browser/playlists_types.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "url/gurl.h"
+
+namespace {
+
+constexpr unsigned int kRetriesCountOnNetworkChange = 1;
+const base::FilePath::StringType kSourceMediaFilesDir(
+    FILE_PATH_LITERAL("source_files"));
+const base::FilePath::StringType kUnifiedMediaFileName(
+    FILE_PATH_LITERAL("media_file"));
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTagForURLLoad() {
+  return net::DefineNetworkTrafficAnnotation("playlists_controller", R"(
+      semantics {
+        sender: "Brave Playlists Controller"
+        description:
+          "Fetching thumbnail image for newly created playlist"
+        trigger:
+          "User-initiated for creating new playlists "
+        data:
+          "Thumbnail for playlist"
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+      })");
+}
+
+void DeleteFile(const base::FilePath& path) {
+  base::DeleteFile(path, true);
+}
+
+bool EnsureDirectoryExists(const base::FilePath& path) {
+  if (base::DirectoryExists(path))
+    return true;
+
+  return base::CreateDirectory(path);
+}
+
+base::FilePath::StringType GetFileNameStringFromIndex(int index) {
+  base::FilePath::StringType file_name =
+#if defined(OS_WIN)
+    base::NumberToString16(index);
+#else
+    base::NumberToString(index);
+#endif
+  return file_name;
+}
+
+size_t WriteToFile(base::File* file, base::StringPiece data) {
+  size_t bytes_written = 0;
+
+  if (file->IsValid()) {
+    if (!data.empty())
+      bytes_written +=
+          std::max(0, file->WriteAtCurrentPos(data.data(), data.size()));
+  }
+
+  return bytes_written;
+}
+
+void AppendToFileThenDelete(const base::FilePath& source_path,
+                            base::File* destination_file,
+                            char* read_buffer,
+                            size_t read_buffer_size) {
+  base::ScopedFILE source_file(base::OpenFile(source_path, "rb"));
+  if (!source_file)
+    return;
```

Comment:
if `OpenFile` fails, we should do more than silently return without notifying anyone

Reply:
Currently, that source file is not included.
but, I think new api `recoverPlaylist` will try again.